rfcs/0030-ensure-index-less-than-length-in-since/0030-ensure-index-less-than-length-in-since.md

```diff
@@ -1 +1,87 @@
-In review, see <https://github.com/nervosnetwork/rfcs/pull/223>
+---
+Number: "0030"
+Category: Consensus (Soft Fork)
+Status: Draft
+Author: Ian Yang
+Organization: Nervos Foundation
+Created: 2021-02-04
+---
+
+# Ensure That Index Is Less Than Length In the Input Since Field Using Epoch With Fraction
+
+## Abstract
+
+This document proposes to fix a bug in the CKB reference implementation. When a transaction input uses the epoch with fraction in the `since` field, the `value` must ensure that the index is less than the length.
+
+## Motivation
+
+The `value` part is an encoded number `epoch + index / length`. In CKB, every block is also assigned an epoch number with a fraction. If the block is the `I`th block in the epoch which epoch number is `E` and length is `L`, its epoch number with the fraction is `E + I / L` and `I` must be less than `L`. 
+
+To verify that a transaction is mature since an absolute epoch number with a fraction, the current implementation assumes that `index` must be less than `length` and checks that:
+
+```
+E > epoch
+or (E == epoch and I / L >= index/length)
+```
+
+This is not consistent when using relative flag in the `since` field, where the `value` is considered as a rational number and added to the cell start epoch.
+
+This document suggests adding a new rule to verify that when `since` uses epoch as the unit, it must ensure that index is less than the length.
+
+See RFC17 [Transaction valid since](../0017-tx-valid-since/0017-tx-valid-since.md) for details of the `since` field encoding patterns.
+
+## Specification
+
+### Since Absolute Epoch
+
+When an input `since` field is present, and
+
+* The `metric_flag` is epoch (01).
+* The `relative_flag` is absolute (0).
+
+Providing that `value` is the encoded number `epoch + index / length`, the transaction is valid only if `index` is less than `length` or both `index` and `length` are zero \*.
```

Comment:
I know this is probably the wrong time to ask for an implementation change, but just for completeness: what is the point of allowing both `index` and `length` as zero? Since this will be converted to `index` as 0 and `length` as 1, what prevents us from removing this rule, and allowing users to fill in `index` as 0, and any arbitrary positive number as `length`?

Reply:
This is also derived from the implementation. I keep it unchanged because it's common in some languages to memset the `since` field to zero and only set the epoch part.